### ISSUE TITLE
Possible to make it work with the torch.compile node?

### ISSUE THREAD

**Issue by Ph0rk0z** [city96/ComfyUI-GGUF#118]
It currently gives an out of range error. Would be nice to get the slight speedup that the inductor or other compile engines can give. Haven't investigated the why super thoroughly yet, but other models do work with it.

**Comment by city96** [comment]
Oh hey, I remember you from those Pascal/FP16 threads about like exl1 or GPTQ lmao. Even feels relevant considering all I get on my volta card with the original BF16 model is a bunch of unsupported dtype errors. So much for comparing that lol.

![image](https://github.com/user-attachments/assets/96299d4a-406c-4cad-9e88-1a5644181921)

Weirdly enough, it seems to be failing [on these lines here](https://github.com/comfyanonymous/ComfyUI/blob/83b01f960ac7c0f4e0c6da539c522c386472b15a/comfy/ldm/flux/model.py#L154-L155), and commenting them out makes it work again except the output is garbage (most likely due to not having actual image embeddings anymore).

Wonder if it just hates the `[...]` indexing notation for some reason.

![image](https://github.com/user-attachments/assets/7a9c89ca-2dd8-455c-af7e-ebf005509eb4)

**Comment by city96** [comment]
Ah yeah, it works if I get rid of the array notation there. Speedup is actually pretty huge, I think it's tracing the decompile kernels as well. I'll test and get a PR for upstream comfy later tonight.

If you wanna try:

```
        img_ids[:, :, 1] = torch.linspace(0, h_len - 1, steps=h_len, device=x.device, dtype=x.dtype).unsqueeze(1)
        img_ids[:, :, 2] = torch.linspace(0, w_len - 1, steps=w_len, device=x.device, dtype=x.dtype).unsqueeze(0)
```

![image](https://github.com/user-attachments/assets/c21640be-69bc-4c19-b1ce-46737c049e15)

**Comment by Ph0rk0z** [comment]
Yes it's me. Now I have 3 gens of GPUs; Ampere, Turning, Pascal. Compiling did work but I only get .0X speedup. On plain SD/XL models I get at least .20s shaved. Unless we find some other compilation engine, I think this is DOA. Maybe it's time to install tensorrt again.

The only thing faster than BF16 is FP8 quants cast to it. Unfortunately output suffers. GGUF Q8 quality is the best, but the speeds :( Especially when I'm using the 2080ti 22g because the 3090s are running LLMs. I am permanently using T5 Q8 though.

I found https://github.com/MinusZoneAI/ComfyUI-Flux1Quantize-MZ and marlin gave the quickest inference to date. Perhaps it's possible to adapt loaded GGUF to use one of the AWQ kernels by shifting the format. He literally just pulled linear out of AWQ and used that. Code from AutoAWQ changes AWQ -> exllama and a similar approach might work. At minimum it shows how to leverage a custom kernel for inference within comfy.

**Comment by simonlui** [comment]
Saving 20 seconds isn't nothing though, as it's around the same amount saved from my experience on an A770 16GB GPU using Intel's Extension for Pytorch custom runtime that allows Intel GPUs to run Pytorch code in the first place. Granted, things do compile a lot faster as it takes less than a minute to compile FLUX in that environment so I think for me, that is good enough of a tradeoff given I don't change my image generation parameters often that would necessitate a rebuild constantly.
I think the `cudagraphs` backend which is what is being used for default for `torch.compile` may not be what you are looking for. But other than `inductor`, it is the only other backend added by ComfyUI for the time being. A suggestion would be to hack the code manually and try using the `tensorrt` backend which is available in mainline Pytorch for testing and see what that does. But I have heard compilation does take quite a while longer with TensorRT so not sure if it would offer anything else substantial here in speedup given that tradeoff.

**Comment by city96** [comment]
I wonder if making the decompilation kernels easier to trace somehow would help, or maybe having the right one be assigned on load to the weight itself instead of having an if/else at runtime. I think the issue is that it exceeds the max depth when trying to trace those parts so it just bails.

That or making those functions be compatible with `torch.jit` (that would be a flat % speedup even without torch compile, possibly? Not sure if it's smart enough to fuse operations but I expect it to be a *bit* faster at least. Last time I tested just slapping on the decorator it failed so that might be a good pointer to why torch.compile also fails.)

**Comment by Ph0rk0z** [comment]
I wouldn't say it failed as much as it didn't bring many gains. There are some warnings when it does compile though. Q8 just needs a slight speedup to match the Fp8 quants, at least without lora. Fusing always helped on LLMs.

**Comment by Ph0rk0z** [comment]
I got it to compile using onediff nexfort but it doesn't seem like a speedup. Maybe I got something wrong?

```
import torch
from onediff.infer_compiler import compile
import os

class TorchCompileModel:
    @classmethod
    def INPUT_TYPES(s):
        return {"required": { "model": ("MODEL",),
                             "backend": (["inductor", "cudagraphs"],),
                              }}
    RETURN_TYPES = ("MODEL",)
    FUNCTION = "patch"

    CATEGORY = "_for_testing"
    EXPERIMENTAL = True

    def patch(self, model, backend):
        os.environ['NEXFORT_FX_FORCE_TRITON_SDPA'] = '1'  #may not be needed
        os.environ['NEXFORT_FUSE_TIMESTEP_EMBEDDING'] = '0' #may not be needed
        #options = '{"mode": "O3"}'  # mode can be O2 or O3
        options= {"mode": "max-optimize:max-autotune:max-autotune", "options": {"inductor.optimize_linear_epilogue": True, "triton.fuse_attention_allow_fp16_reduction": True}}
        m = model.clone()
        #m.add_object_patch("diffusion_model", torch.compile(model=m.get_model_object("diffusion_model"), backend=backend))
        m.add_object_patch("diffusion_model", compile(m.get_model_object("diffusion_model"), backend="nexfort", options=options))
        return (m, )

NODE_CLASS_MAPPINGS = {
    "TorchCompileModel": TorchCompileModel,
}
```

**Comment by Ph0rk0z** [comment]
Sadly 4xxx cards are out of my reach.

**Comment by StrongerXi** [comment]
#243 addresses this.

**Comment by Heliumrich** [comment]
Strangely I can't make torch.compile work with Qwen-Image Q4_K_M when using "Qwen-lighting 8steps" lora 😔 
The LoRA is FP32 so maybe that's why... (I tried to resave it as BF16 except the INT64 parts, and same "recompilation targets reached" error)

Everything is updated, ComfyUI nightly.

I've commented here: https://github.com/city96/ComfyUI-GGUF/issues/310#issuecomment-3177341057
With two links to issues I've opened on KJ nodes and ComfyUI main repo

I believe either GGUF or ComfyUI is causing issue with that model + torch.compile when using any lora

**Comment by StrongerXi** [comment]
Regarding GGUF + LoRA recompilation issue, see my comment in https://github.com/kijai/ComfyUI-KJNodes/issues/363#issuecomment-3180341161.
